File: sdks/client/python/lenny/stream.py
# SPDX-License-Identifier: MIT

"""The section 15.1 Server-Sent Events session-event stream.

:class:`EventStream` consumes ``GET /v1/sessions/{id}/events``, the
section 15.1 SSE stream of session activity. It decodes each SSE frame
into a :class:`StreamEvent`, reconnects on a transport disconnect, and
resumes from the last delivered sequence with the ``Last-Event-ID``
header so the section 15.1 streaming-reconnect contract holds: the
backlog is replayed, no event is skipped, and no event is delivered
twice.

The transport uses only the Python standard library
(:mod:`urllib.request`). :class:`EventStream` is the synchronous
iterator; :class:`AsyncEventStream` is the ``async``/``await`` form,
which runs each blocking read on a worker thread so an ``async for``
caller does not block the event loop.
"""

from __future__ import annotations

import http.client
import json
import random
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass, field
from typing import Any, AsyncIterator, Iterator, Optional

from .auth import Authenticator
from .errors import APIError
from .retry import DEFAULT_RETRY_POLICY, RetryPolicy

#: Identifies the SDK in the ``User-Agent`` request header.
_STREAM_USER_AGENT = "lenny-client-sdk-python"


@dataclass
class StreamEvent:
    """One decoded frame from the section 15.1 SSE event stream."""

    #: The monotonic per-session sequence the gateway assigns. It is the
    #: cursor a reconnect resumes from: the SDK sends the last delivered
    #: ``seq`` as the ``Last-Event-ID`` header. A frame with no ``id:``
    #: field has ``seq`` 0.
    seq: int = 0

    #: The SSE event type (the ``event:`` field), for example
    #: ``message_delivered``, ``response``, or ``state_changed``. Empty
    #: when the frame carried no ``event:`` field.
    type: str = ""

    #: The event payload, the JSON text carried in the ``data:`` field.
    data: str = ""

    def json(self) -> Any:
        """Decode :attr:`data` as JSON and return the parsed value."""
        return json.loads(self.data) if self.data else None


@dataclass
class StreamOptions:
    """Settings for an :meth:`~lenny.client.Client.stream_events` call.

    Attributes:
        last_event_id: The sequence the stream resumes after. Zero (the
            default) starts from the beginning of the retained backlog.
            A non-zero value sends the section 15.1 ``Last-Event-ID``
            header on the initial request so the gateway replays only
            events after that cursor.
    """

    last_event_id: int = 0


@dataclass
class _StreamConfig:
    """The transport surface the stream loop reads.

    It is the subset of the :class:`~lenny.client.Client` internals the
    stream needs, passed in so the stream module does not reach into the
    client's private state.
    """

    base_url: str
    timeout: float
    retry: RetryPolicy
    auth: Optional[Authenticator] = None
    tenant_id: str = ""
    rng: random.Random = field(default_factory=random.Random)


class _SSEParser:
    """Accumulates the lines of an SSE byte stream into frames.

    The parser holds the ``id``, ``event``, and ``data`` fields of one
    frame and emits a :class:`StreamEvent` on the blank line that
    terminates the frame.
    """

    def __init__(self) -> None:
        self._seq = 0
        self._type = ""
        self._data: list[str] = []
        self._started = False

    def push_line(self, line: str) -> Optional[StreamEvent]:
        """Feed one SSE line.

        Return the completed :class:`StreamEvent` when ``line`` is the
        blank line that terminates a frame, and ``None`` otherwise.
        """
        if line == "":
            # A blank line terminates the frame. Dispatch it only when a
            # recognized field was seen, so a blank line that follows
            # only comment lines or blank padding is skipped rather than
            # dispatched as an empty frame.
            if not self._started:
                return None
            event = StreamEvent(
                seq=self._seq,
                type=self._type,
                data="\n".join(self._data),
            )
            self._reset()
            return event

        field_name, value = _split_sse_line(line)
        if field_name == "id":
            self._started = True
            if value.isdigit():
                self._seq = int(value)
        elif field_name == "event":
            self._started = True
            self._type = value
        elif field_name == "data":
            self._started = True
            # The SSE format joins multiple data lines with a newline.
            self._data.append(value)
        # A comment line (an empty field name, the line started with a
        # colon) or an unknown field name contributes nothing.
        return None

    def _reset(self) -> None:
        self._seq = 0
        self._type = ""
        self._data = []
        self._started = False


def _split_sse_line(line: str) -> tuple[str, str]:
    """Split one SSE line into its field name and value.

    The SSE format separates them with the first colon and strips a
    single leading space from the value. A line with no colon is a field
    name with an empty value.
    """
    idx = line.find(":")
    if idx < 0:
        return line, ""
    value = line[idx + 1 :]
    if value.startswith(" "):
        value = value[1:]
    return line[:idx], value


class EventStream:
    """Synchronous iterator over the section 15.1 SSE event stream.

    Iterate it with a ``for`` loop. The iterator reconnects on a
    transport disconnect, resuming from the ``Last-Event-ID`` cursor of
    the last delivered event, and drops any replayed event at or below
    that cursor so no event is skipped or delivered twice.

    The gateway holds the SSE connection open until the client
    disconnects; it does not close a live stream on its own. The SDK
    therefore treats any connection close, including a clean end of
    body, as an unexpected disconnect and reconnects, the same reconnect
    behavior the WHATWG EventSource model specifies. A reconnect that
    delivers an event resets the attempt counter, so a long-lived stream
    reconnects indefinitely; consecutive reconnects that deliver no
    event are bounded by the retry policy's ``max_attempts``.

    Iteration ends when :meth:`close` is called (a clean
    caller-requested stop), when the gateway returns a non-retryable
    HTTP status (the iterator raises the typed
    :class:`~lenny.errors.APIError`), or when consecutive zero-progress
    reconnects exhaust the retry budget. A retryable HTTP status (429,
    5xx) is retried like a transport disconnect.

    The stream is a context manager: a ``with`` block closes it on exit.
    """

    def __init__(
        self,
        config: _StreamConfig,
        session_id: str,
        options: Optional[StreamOptions] = None,
    ) -> None:
        self._config = config
        self._path = "/v1/sessions/" + urllib.parse.quote(session_id, safe="") + "/events"
        opt = options or StreamOptions()
        self._last_seq = opt.last_event_id if opt.last_event_id > 0 else 0
        self._closed = False
        self._response: Optional[http.client.HTTPResponse] = None

    def __iter__(self) -> Iterator[StreamEvent]:
        return self._run()

    def __enter__(self) -> "EventStream":
        return self

    def __exit__(self, *_exc: Any) -> None:
        self.close()

    def close(self) -> None:
        """Stop the stream.

        It marks the stream closed and tears down any open connection.
        An iterator blocked on a read returns once the underlying socket
        is closed. :meth:`close` is idempotent.
        """
        self._closed = True
        self._close_response()

    def _close_response(self) -> None:
        response = self._response
        self._response = None
        if response is not None:
            try:
                response.close()
            except OSError:
                # The connection is already torn down; nothing to free.
                pass

    def _run(self) -> Iterator[StreamEvent]:
        policy = self._config.retry
        # failures counts consecutive reconnects that delivered no
        # event. It bounds reconnection against a permanently broken
        # stream and spaces attempts with the retry policy's backoff. A
        # connection that delivers an event resets it to zero.
        failures = 0
        while not self._closed:
            if failures > 0:
                delay = policy.delay_for_attempt(failures, self._config.rng)
                if _sleep_or_closed(delay, self._is_closed):
                    return

            try:
                response = self._open()
            except APIError as exc:
                # A non-retryable HTTP status ends the stream with the
                # typed error; a retryable status reconnects.
                if not exc.retryable:
                    raise
                failures += 1
                if failures >= policy.max_attempts:
                    raise
                continue
            except OSError:
                # A transport-level failure before any response
                # (connection refused, DNS, a reset). Reconnect.
                failures += 1
                if failures >= policy.max_attempts:
                    raise
                continue

            progressed = False
            read_error: Optional[BaseException] = None
            try:
                for event in self._read_frames(response):
                    # Drop a replayed event at or below the last
                    # delivered sequence so a reconnect does not surface
                    # a duplicate.
                    if event.seq != 0 and event.seq <= self._last_seq:
                        continue
                    yield event
                    if event.seq > self._last_seq:
                        self._last_seq = event.seq
                        progressed = True
            except OSError as exc:
                # A read error mid-stream. Reconnect from the last
                # delivered cursor.
                read_error = exc
            finally:
                self._close_response()

            if self._closed:
                return
            # A transport disconnect, a clean end of body, or a read
            # error. Reconnect. A connection that delivered an event
            # restarts the attempt counter; otherwise the counter
            # advances toward the max_attempts ceiling.
            if progressed:
                failures = 0
                continue
            failures += 1
            if failures >= policy.max_attempts:
                # The stream made no progress across the full retry
                # budget. Surface the last read failure, or stop
                # cleanly when the connection kept ending at a clean
                # end of body.
                if read_error is not None:
                    raise read_error
                return

    def _is_closed(self) -> bool:
        return self._closed

    def _open(self) -> http.client.HTTPResponse:
        """Open one SSE connection.

        Return the streaming response on a 2xx. A non-2xx status raises
        the typed :class:`~lenny.errors.APIError`; a transport failure
        raises an :class:`OSError`.
        """
        headers: dict[str, str] = {
            "Accept": "text/event-stream",
            "User-Agent": _STREAM_USER_AGENT,
        }
        if self._last_seq > 0:
            # The section 15.1 reconnect cursor. The gateway replays the
            # retained backlog with ``seq`` greater than this value
            # before live delivery resumes.
            headers["Last-Event-ID"] = str(self._last_seq)
        if self._config.tenant_id:
            headers["X-Lenny-Tenant-ID"] = self._config.tenant_id
        if self._config.auth is not None:
            self._config.auth.apply(headers)

        request = urllib.request.Request(
            self._config.base_url + self._path,
            headers=headers,
            method="GET",
        )
        try:
            # urlopen is typed to return Any; for an HTTP(S) URL the
            # concrete return is an http.client.HTTPResponse, whose
            # incremental readline the SSE parser drives.
            opened = urllib.request.urlopen(request, timeout=self._config.timeout)
            response: http.client.HTTPResponse = opened
        except urllib.error.HTTPError as exc:
            # A non-2xx response is a complete round trip; decode the
            # section 15.1 envelope and surface the typed error.
            body = exc.read()
            exc.close()
            raise APIError.from_response(exc.code, body) from None
        except urllib.error.URLError as exc:
            raise OSError(f"lenny: stream connection failed: {exc.reason}") from exc

        self._response = response
        return response

    def _read_frames(
        self, response: http.client.HTTPResponse
    ) -> Iterator[StreamEvent]:
        """Decode the SSE byte stream of one connection into events.

        The generator returns when the body ends. A frame buffered
        without its terminating blank line is discarded.
        """
        parser = _SSEParser()
        while True:
            if self._closed:
                return
            raw = response.readline()
            if raw == b"":
                # The body ended. A clean stream ends on a frame
                # boundary; a partial frame in the parser is discarded.
                return
            # SSE lines are LF-terminated; tolerate a CRLF stream by
            # trimming a trailing CR.
            line = raw.rstrip(b"\n").rstrip(b"\r").decode("utf-8", "replace")
            event = parser.push_line(line)
            if event is not None:
                yield event


class AsyncEventStream:
    """Asynchronous iterator over the section 15.1 SSE event stream.

    It is the ``async``/``await`` counterpart of :class:`EventStream`.
    Each blocking connection and frame read runs on a worker thread, so
    an ``async for`` caller does not block the event loop. The reconnect
    and dedup behavior is identical to :class:`EventStream`.
    """

    def __init__(
        self,
        config: _StreamConfig,
        session_id: str,
        options: Optional[StreamOptions] = None,
    ) -> None:
        self._stream = EventStream(config, session_id, options)

    def __aiter__(self) -> AsyncIterator[StreamEvent]:
        return self._run()

    async def __aenter__(self) -> "AsyncEventStream":
        return self

    async def __aexit__(self, *_exc: Any) -> None:
        await self.aclose()

    async def aclose(self) -> None:
        """Stop the stream.

        It closes the underlying synchronous stream so a worker-thread
        read returns. :meth:`aclose` is idempotent.
        """
        self._stream.close()

    async def _run(self) -> AsyncIterator[StreamEvent]:
        import asyncio

        loop = asyncio.get_running_loop()
        # The synchronous iterator is advanced one event at a time on a
        # worker thread. _NO_MORE is the sentinel the worker returns
        # when the synchronous iterator is exhausted.
        iterator = iter(self._stream)

        def next_event() -> Any:
            try:
                return next(iterator)
            except StopIteration:
                return _NO_MORE

        try:
            while True:
                event = await loop.run_in_executor(None, next_event)
                if event is _NO_MORE:
                    return
                yield event
        finally:
            self._stream.close()


#: Sentinel marking the synchronous iterator as exhausted, returned by
#: the worker-thread step in :class:`AsyncEventStream`.
_NO_MORE = object()


def _sleep_or_closed(delay: float, is_closed: Any) -> bool:
    """Wait ``delay`` seconds, or return early when the stream closes.

    Return ``True`` when the stream was closed during the wait so the
    caller stops, and ``False`` when the full delay elapsed.
    """
    import time

    if delay <= 0:
        return bool(is_closed())
    deadline = time.monotonic() + delay
    while time.monotonic() < deadline:
        if is_closed():
            return True
        time.sleep(min(0.02, deadline - time.monotonic()))
    return bool(is_closed())

File: sdks/client/go/lenny/stream.go
// SPDX-License-Identifier: MIT

package lenny

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
)

// StreamEvent is one decoded frame from the §15.1
// GET /v1/sessions/{id}/events Server-Sent Events stream.
type StreamEvent struct {
	// Seq is the monotonic per-session sequence the gateway assigns.
	// It is the cursor a reconnect resumes from: the SDK sends the
	// last delivered Seq as the Last-Event-ID header.
	Seq uint64

	// Type is the SSE event type (the `event:` field), for example
	// message_delivered, response, or state_changed.
	Type string

	// Data is the event payload, the JSON object carried in the
	// `data:` field.
	Data json.RawMessage
}

// StreamOptions configures a StreamEvents call.
type StreamOptions struct {
	// LastEventID is the sequence the stream resumes after. Zero
	// starts from the beginning of the retained backlog. A non-zero
	// value sends the §15.1 Last-Event-ID header on the initial
	// request so the gateway replays only events after that cursor.
	LastEventID uint64
}

// Stream is a live handle to a §15.1 session event stream. Construct
// it with Client.StreamEvents. The caller ranges over Events until
// the channel closes, then calls Err to learn why the stream ended.
//
// A Stream is not safe for concurrent use; a single goroutine reads
// Events and then reads Err.
type Stream struct {
	events chan StreamEvent
	done   chan struct{}
	err    error
}

// Events returns the channel of decoded events. The channel is
// closed when the stream ends: on context cancellation, on a
// non-retryable HTTP status, or when the gateway closes the stream
// without a transport error. After the channel closes, call Err.
func (s *Stream) Events() <-chan StreamEvent { return s.events }

// Err returns the error that ended the stream. It returns nil when
// the context was canceled (a clean caller-requested stop) or when
// the gateway closed the stream cleanly. Err must be read only after
// the Events channel is observed closed.
func (s *Stream) Err() error {
	<-s.done
	return s.err
}

// StreamEvents opens the §15.1 GET /v1/sessions/{id}/events
// Server-Sent Events stream for a session and returns a Stream that
// delivers decoded events on a channel.
//
// On a transport disconnect the Stream reconnects automatically,
// sending the last delivered sequence as the Last-Event-ID header so
// the gateway replays the retained backlog from that cursor. Any
// replayed event whose Seq is at or below the last delivered Seq is
// dropped, so the §15.1 reconnect contract holds: the backlog is
// replayed, no event is skipped, and no event is delivered twice.
// Reconnect attempts are spaced by the Client retry policy's backoff.
//
// The stream stops when ctx is canceled (Err then reports nil), when
// the gateway returns a non-retryable HTTP status (Err reports the
// typed APIError), or when the gateway closes the stream without a
// transport error. A retryable HTTP status (429, 5xx) on the initial
// request or on a reconnect is retried like a transport disconnect.
func (c *Client) StreamEvents(ctx context.Context, sessionID string, opt StreamOptions) *Stream {
	s := &Stream{
		events: make(chan StreamEvent),
		done:   make(chan struct{}),
	}
	go c.runStream(ctx, sessionID, opt.LastEventID, s)
	return s
}

// runStream is the Stream's background loop. It opens the SSE
// connection, relays decoded events, and reconnects on a transport
// disconnect until ctx is canceled or a non-retryable status ends
// the stream.
//
// The gateway holds the SSE connection open until the client
// disconnects; it does not close a live stream on its own. The SDK
// therefore treats any connection close, including a clean EOF, as
// an unexpected disconnect and reconnects, the same reconnect
// behavior the WHATWG EventSource model specifies. A reconnect that
// delivers no further event is retried up to the retry policy's
// MaxAttempts so a permanently unavailable stream does not loop
// forever; a reconnect that delivers at least one event resets the
// attempt counter, so a long-lived stream reconnects indefinitely.
func (c *Client) runStream(ctx context.Context, sessionID string, lastSeq uint64, s *Stream) {
	defer close(s.events)
	defer close(s.done)

	path := "/v1/sessions/" + url.PathEscape(sessionID) + "/events"
	policy := c.retry
	// failures counts consecutive reconnects that delivered no event.
	// It bounds reconnection against a permanently broken stream and
	// spaces attempts with the retry policy's backoff. A connection
	// that delivers an event resets it to zero.
	failures := 0

	for {
		if err := ctx.Err(); err != nil {
			return
		}
		if failures > 0 {
			delay := policy.delayForAttempt(failures, c.rng)
			if err := sleepCtx(ctx, delay); err != nil {
				return
			}
		}

		delivered, retErr := c.streamOnce(ctx, path, lastSeq, s)
		progressed := delivered > lastSeq
		if delivered > lastSeq {
			lastSeq = delivered
		}

		switch {
		case ctx.Err() != nil:
			// The caller canceled the context. A clean stop.
			return
		case retErr != nil && retErr.permanent:
			s.err = retErr.err
			return
		default:
			// A transport disconnect, a clean EOF, or a retryable
			// status. Reconnect. A connection that delivered an event
			// restarts the attempt counter; otherwise the counter
			// advances toward the MaxAttempts ceiling.
			if progressed {
				failures = 0
				continue
			}
			failures++
			if failures >= policy.MaxAttempts {
				// The stream made no progress across the full retry
				// budget. Surface the last failure, or nil when the
				// connection kept ending at a clean EOF.
				if retErr != nil {
					s.err = retErr.err
				}
				return
			}
		}
	}
}

// streamError carries the outcome of one streamOnce connection. A
// permanent error ends the stream; a non-permanent error triggers a
// reconnect. A nil streamError means the connection ended at a clean
// EOF, which the reconnect loop also treats as a disconnect.
type streamError struct {
	err       error
	permanent bool
}

// streamOnce opens one SSE connection, relays every frame with Seq
// greater than afterSeq onto the Stream channel, and returns the
// highest Seq it delivered together with the reason the connection
// ended. A nil return error means the gateway closed the stream
// cleanly.
func (c *Client) streamOnce(ctx context.Context, path string, afterSeq uint64, s *Stream) (uint64, *streamError) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, c.baseURL+path, nil)
	if err != nil {
		return afterSeq, &streamError{err: fmt.Errorf("lenny: build stream request: %w", err), permanent: true}
	}
	req.Header.Set("Accept", "text/event-stream")
	req.Header.Set("User-Agent", userAgent)
	if afterSeq > 0 {
		// The §15.1 reconnect cursor. The gateway replays the retained
		// backlog with Seq greater than this value before live
		// delivery resumes.
		req.Header.Set("Last-Event-ID", strconv.FormatUint(afterSeq, 10))
	}
	if c.tenantID != "" {
		req.Header.Set("X-Lenny-Tenant-ID", c.tenantID)
	}
	if c.auth != nil {
		if err := c.auth.Apply(req); err != nil {
			return afterSeq, &streamError{err: fmt.Errorf("lenny: apply auth: %w", err), permanent: true}
		}
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		// A transport-level failure (connection refused, reset
		// mid-stream, DNS). Reconnect.
		return afterSeq, &streamError{err: err}
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		body, _ := io.ReadAll(resp.Body)
		apiErr := decodeAPIError(resp.StatusCode, body)
		// A retryable status (429, 5xx) is treated like a transport
		// disconnect: reconnect after a backoff. Any other status
		// ends the stream and surfaces the typed error.
		return afterSeq, &streamError{err: apiErr, permanent: !apiErr.Retryable}
	}

	delivered := afterSeq
	parser := newSSEParser(resp.Body)
	for {
		ev, err := parser.next()
		if err != nil {
			if err == io.EOF {
				// The gateway closed the stream. With a live context
				// this is a clean end; the caller's runStream decides
				// whether to reconnect.
				return delivered, nil
			}
			// A read error mid-stream. Reconnect from the last
			// delivered cursor.
			return delivered, &streamError{err: err}
		}
		// Drop a replayed event at or below the last delivered
		// sequence so a reconnect does not surface a duplicate.
		if ev.Seq != 0 && ev.Seq <= delivered {
			continue
		}
		select {
		case <-ctx.Done():
			return delivered, &streamError{err: ctx.Err()}
		case s.events <- ev:
			if ev.Seq > delivered {
				delivered = ev.Seq
			}
		}
	}
}

// sseParser decodes a Server-Sent Events byte stream into
// StreamEvents. It accumulates the id, event, and data fields of one
// frame and emits a StreamEvent when the blank line that terminates
// the frame is read.
type sseParser struct {
	scanner *bufio.Scanner
}

// newSSEParser returns an sseParser reading from r. The scanner
// buffer is sized for large data payloads.
func newSSEParser(r io.Reader) *sseParser {
	sc := bufio.NewScanner(r)
	sc.Buffer(make([]byte, 0, 64*1024), 4<<20)
	return &sseParser{scanner: sc}
}

// next returns the next complete SSE frame. It returns io.EOF when
// the stream ends with no partial frame buffered, or a non-EOF error
// when the underlying reader fails.
func (p *sseParser) next() (StreamEvent, error) {
	var (
		ev      StreamEvent
		data    strings.Builder
		started bool
	)
	for p.scanner.Scan() {
		line := p.scanner.Text()
		if line == "" {
			// A blank line terminates the frame. Dispatch it only when
			// a recognized field was seen, so a blank line that follows
			// only comment lines or blank padding is skipped rather
			// than dispatched as an empty frame.
			if !started {
				continue
			}
			ev.Data = json.RawMessage(data.String())
			return ev, nil
		}
		field, value := splitSSELine(line)
		switch field {
		case "id":
			started = true
			if n, err := strconv.ParseUint(value, 10, 64); err == nil {
				ev.Seq = n
			}
		case "event":
			started = true
			ev.Type = value
		case "data":
			started = true
			// The SSE format joins multiple data lines with a newline.
			if data.Len() > 0 {
				data.WriteByte('\n')
			}
			data.WriteString(value)
		default:
			// A comment line (an empty field name, the line started
			// with a colon) or an unknown field name. The SSE format
			// permits both; neither contributes to the frame.
		}
	}
	if err := p.scanner.Err(); err != nil {
		return StreamEvent{}, err
	}
	// The reader reached EOF. A frame buffered without its blank-line
	// terminator is discarded; a clean stream ends on a frame
	// boundary.
	return StreamEvent{}, io.EOF
}

// splitSSELine splits one SSE line into its field name and value.
// The SSE format separates them with the first colon and strips a
// single leading space from the value. A line with no colon is a
// field name with an empty value.
func splitSSELine(line string) (field, value string) {
	idx := strings.IndexByte(line, ':')
	if idx < 0 {
		return line, ""
	}
	field = line[:idx]
	value = line[idx+1:]
	value = strings.TrimPrefix(value, " ")
	return field, value
}
